core/src/sumeragi/fault.rs

```diff
@@ -39,865 +37,1333 @@ impl FaultInjection for NoFault {
-    for SumeragiWithFault<G, F>
-{
-    type Result = ();
-
-    async fn handle(&mut self, receipt_timeout: CheckReceiptTimeout) {
-        if self
-            .txs_awaiting_receipts
-            .contains_key(&receipt_timeout.tx_hash)
-        {
-            self.vote_for_view_change(receipt_timeout.proof).await;
+    fn try_get_online_topology(&self, network_topology: &Topology) -> Result<Topology> {
+        use crate::sumeragi::network_topology::GenesisBuilder;
+        let (online_peers, offline_peers) =
+            self.check_peers_status(&self.peer_id, network_topology);
+        let set_a_len = network_topology.min_votes_for_commit();
+        if online_peers.len() < set_a_len {
+            return Err(eyre!("Not enough online peers for consensus."));
         }
+        let genesis_topology = if network_topology.sorted_peers().len() == 1 {
+            network_topology.clone()
+        } else {
+            let set_a: HashSet<_> = online_peers[..set_a_len].iter().cloned().collect();
+            let set_b: HashSet<_> = online_peers[set_a_len..]
+                .iter()
+                .cloned()
+                .chain(offline_peers.into_iter())
+                .collect();
+            #[allow(clippy::expect_used)]
+            GenesisBuilder::new()
+                .with_leader(self.peer_id.clone())
+                .with_set_a(set_a)
+                .with_set_b(set_b)
+                .build()
+                .expect("Preconditions should be already checked.")
+        };
+        iroha_logger::info!("Waiting for active peers finished.");
+        Ok(genesis_topology)
     }
-}
 
-#[async_trait::async_trait]
-impl<G: GenesisNetworkTrait, F: FaultInjection> Handler<CheckCreationTimeout>
-    for SumeragiWithFault<G, F>
-{
-    type Result = ();
+    pub fn wait_for_peers(&self, network_topology: &Topology) {
+        iroha_logger::info!("Waiting for active peers",);
+        for i in 0..120 {
+            self.connect_peers(&network_topology);
 
-    async fn handle(&mut self, creation_timeout: CheckCreationTimeout) {
-        if self
-            .txs_awaiting_created_block
-            .contains(&creation_timeout.tx_hash)
-        {
-            self.vote_for_view_change(creation_timeout.proof).await;
+            let online_peers = self.get_online_peer_keys();
+
+            let peer_count_required = network_topology.min_votes_for_commit();
+
+            if online_peers.len() + 1 >= peer_count_required {
+                return;
+            }
+
+            let reconnect_in_ms = 250;
+            std::thread::sleep(Duration::from_millis(reconnect_in_ms));
         }
+        panic!("Timed out waiting for peers to come online");
+    }
+
+    pub fn pipeline_time(&self) -> Duration {
+        self.block_time + self.commit_time
     }
 }
 
-#[async_trait::async_trait]
-impl<G: GenesisNetworkTrait, F: FaultInjection> Handler<CheckCommitTimeout>
-    for SumeragiWithFault<G, F>
+fn block_commit<F>(
```

Comment:
Why not method? the same question for other functions with signature `fn(&SumeragiWithFault<F>, ...)`

Reply:
To me it's about frame of mind. This function is not associated with some object. It just does stuff, in fact it does things to all the things passed in. It commits the block, it communicates with the outside world through `sumeragi` and it modifies the state machine data. All of these three are equal participants and so to me it doesn't make sense to make this function method when it could be a method on all three.